main/reference/vstorage-ref.md

```diff
@@ -1,13 +1,83 @@
-# VStorage Reference
+# VStorage in Agoric
 
-See also:
+In the Agoric platform, VStorage, or a Virtual Storage, is a key-value store that:
 
-- [Querying VStorage](/guides/getting-started/contract-rpc#querying-vstorage)
-- [Publishing to chainStorage](../guides/zoe/pub-to-storage)
-- [x/vstorage module](https://github.com/Agoric/agoric-sdk/tree/003f0c2232815a8d64a3f9a5b05521a10160ce34/golang/cosmos/x/vstorage#readme)
+- provides a **read-only interface for clients of the consensus layer**. Clients can query the stored data using specific paths.
+- is organized in a **hierarchical, path-based structure**. Each fragment of data is stored at a specific path, and clients can query these paths to retrieve the data.
+- can be written through a specialized API called **[chainStorage](https://docs.agoric.com/guides/zoe/pub-to-storage.html#publishing-to-chainstorage)**. This API allows the VM to manage data storage in a controlled and secure manner.
+
+![vstorage query diagram](../guides/getting-started/assets/vstorage-brand-q.svg)
+
+## VStorage Hierarchy
+
+VStorage is structured as a tree with paths and nodes that store the actual data. This design facilitates easy querying while ensuring data consistency and security.
+
+For example, here is an example of [Inter Protocol key structure](https://github.com/Agoric/agoric-sdk/tree/agoric-upgrade-13/packages/inter-protocol#reading-data-off-chain) inside Agoric:
+
+```
+- published
+    - vaultFactory
+        - governance
+        - metrics
+        - managers
+            - manager0
+                - metrics
+                - governance
+                - vaults
+                    - vault0
+```
+
+You can then use this hierarchy to access the data fragments.
+```sh
+# lists vaults
+$ agd query vstorage keys 'published.vaultFactory.managers.manager0.vaults'
+children:
+- vault0
+```
```

Comment:
This is good.

Reply:
It assumes knowledge of the `agd` CLI. Perhaps add a cross-reference to https://docs.agoric.com/guides/agoric-cli/agd-query-tx.html#query-commands ?

Also, add a section about `agd query vstorage` in https://docs.agoric.com/guides/agoric-cli/agd-query-tx.html ?

Reply:
Do we expect our readers to be more familiar with `agd` or with JS APIs?

https://docs.agoric.com/guides/getting-started/ui-tutorial/querying-vstorage.html has an example using `chainStorageWatcher.watchLatest` to access `published.agoricNames.instance`. Did you consider using that?

Reply:
I understand. I have added both, JS APIs and `agd` here.

Reply:
Added here #1122

Reply:
The JS API that has been added is how to _write_ to vstorage / chainstorage from a contract. I was suggesting adding the API for _querying_ vstorage from a client dapp, as suggested in #947. I suppose both are relevant. I'm not sure how big the scope should grow here, though.